sys/net/gnrc/routing/aodvv2/aodv.c

```diff
@@ -0,0 +1,331 @@
+                        .hoplimit = AODVV2_MAX_HOPCOUNT,
+                        .metricType = aodv_metric_type,
+                        .origNode = (struct node_data) {
+                            .addr = aodvv2_netaddr_local,
+                            .metric = 0,
+                            .seqnum = seqnum,
+                        },
+                        .targNode = (struct node_data) {
+                            .addr = netaddr_dest,
+                        },
+                        .timestamp = (timex_t) {0,0} /* never used, exists merely
+                                                      * to silence compiler*/
+                    };
+
+                    /* since we're originating the RREQ, add it to the RREQ
+                     * table or add its predecessor (if any) */
+                    rreqtable_is_redundant(&new_rreq_data);
+
+                    aodv_packet_writer_send_rreq(&new_rreq_data,
+                                                 &aodvv2_netaddr_mcast);
+                }
+                else {
+                    /* Reply to the FIB so that it can stop blocking */
+                    msg_reply(&msg, &msg);
+                }
+                break;
+            }
+            default:
+                DEBUG("[aodvv2] received msg of unexpected type: %d\n", msg.type);
+                break;
+        }
+    }
+
+    return NULL;
+}
+
+/**
+ * Send finalized RFC 5444 packets. This callback is called by every
+ * aodv_packet_writer_send_* function.
+ */
+static void aodv_send_packet(struct rfc5444_writer *wr __attribute__ ((unused)),
+                          struct rfc5444_writer_target *iface __attribute__((unused)),
+                          void *buffer, size_t length)
+{
+    ipv6_addr_t addr_send;
+    gnrc_pktsnip_t *payload, *pkt_with_udp, *pkt_with_ip;
+
+    /* fetch the address the packet is supposed to be sent to (i.e. to a
+     * specific node or the multicast address) from the writer_target struct
+     * iface* is stored in. This is a bit hacky, but it does the trick. */
+    struct writer_target *wt = container_of(iface, struct writer_target, interface);
+
+    print_json_pkt_sent(wt);
+    netaddr_to_ipv6_addr_t(&wt->target_addr, &addr_send);
+
+    /* allocate payload (i.e. the RFC5444 packet)*/
+    payload = gnrc_pktbuf_add(NULL, buffer, length, GNRC_NETTYPE_UNDEF);
+    if (payload == NULL) {
+        DEBUG("Error: unable to copy data to packet buffer\n");
+        return;
+    }
+    /* allocate UDP header, set payload, set source port := destination port */
+    pkt_with_udp = gnrc_udp_hdr_build(payload, MANET_PORT, MANET_PORT);
+    if (pkt_with_udp == NULL) {
+        DEBUG("Error: unable to allocate UDP header\n");
+        gnrc_pktbuf_release(payload);
+        return;
+    }
+    /* allocate IPv6 header, set pkt_with_udp as payload */
+    pkt_with_ip = gnrc_ipv6_hdr_build(pkt_with_udp, &aodv_ipv6_addr_local,
+                                      &addr_send);
```

Comment:
> RREPs to have the wrong source & destination addresses in the IP header (!).

How wrong are they? If they are still sensible addresses I guess you are setting the wrong addresses here. If they are total garbage you might overwrite something accidentaly. My choice of tool for this would be `gdb`:

```gdb
watch *<address of first byte in destination or source address>
```